Alternate Approaches To Context Hashing

danieldresser-ie opened the pull request:
This is some exploration of things we can do to reduce the cost of hashing contexts, all based around the idea that a sum of the hashes of ( entry name, entry value ) is sufficient to uniquely represent the context ( since we do not care about order of context entries ).

There are 3 versions here:

V1:  store a running sum of the hash of all current context entries.  When an entry is set, compute the hash of the current value, subtract it off, the add the hash of the incoming value.
V2:  store a running sum of the hash of all current context entries, plus the hash of each entry.  When an entry is set, subtract off the stored hash, the add the hash of the incoming value.
V3: store the hash of each entry.  To compute the current hash, run through all entries and sum their hashes

These versions are kinda in reverse order - V1 is the version we discussed first, but it actually makes the most changes to the current code.  The diffs to master are more meaningful than diffing V2 and V3 from the previous commit, but this seemed like the quickest way to share all 3.

My preliminary performance results are below.  The basic summary is that V2 seems strictly inferior to V3 - if we're going to use the extra storage, summing the hashes when you need a total hash is insignificant in expense, and getting rid of the running accumulator makes the code much simpler, and harder to introduce bizarre inconsistent states.

So the main question is whether the extra storage cost is worth the reduction in complexity from V1 to V3.  I'm leaning towards thinking it is, particularly because you have some schemes for reducing the number of times we need to allocate cache entries, which could make V1 a lot harder to get working, but could also reduce the already farily small extra storage cost of V3.

Here are performance results, comparing 5 versions of the code, from current master to the V4 commit, with and without this Cortex PR: https://github.com/ImageEngine/cortex/pull/1130, and with the context initialized with a small state of 10 strings with 10 characters, or a larger state with 50 strings with 20 characters:

| --- | testContextHashPerformance | testContextHashPerformanceStartInitialized |
| --- | --- | --- |
| **default IECore, 10 existing 10 character strings** | - | - |
| original | 0.76s | 0.71s |
| V1 | 0.34s | 0.32s |
| V2 | 0.38s | 0.31s |
| V3 | 0.40s | 0.32s |
| V4 | 0.46s | 0.39s |
| **inlined MurmurHash, 10 existing 10 character strings** | - | - |
| original | 0.73s | 0.69s |
| V1 | 0.33s | 0.30s |
| V2 | 0.35s | 0.30s |
| V3 | 0.35s | 0.30s |
| V4 | 0.41s | 0.37s |
| **inlined MurmurHash, 50 existing 20 character strings** | - | - |
| original | 2.22s | 2.13s |
| V1 | 0.52s | 0.44s |
| V2 | 0.59s | 0.46s |
| V3 | 0.62s | 0.46s |
| V4 | 0.87s | 0.74s |

themissingcow commented:
Interesting idea @danieldresser-ie! Please excuse my massively newbie q, does this increase the chances of a collision?

danieldresser-ie commented:
Whether this affects the chances of collision is a crucial question - I've talked about this with John a bit, but I should certainly record my thoughts here.  I would be happier if I had a slightly stronger answer.  I wish I could find more examination of this question in the literature, but haven't turned up much.  The strongest suggestion that other people think it's fine is probably that Java uses a "sum-of-hashes" for hashing sets, so they apparently think it's fine, though they don't justify it particularly, and they don't have the strong requirement we do that hashes never collide ( in the infinitesimally rare case of the Gaffer hash collision, this causes undefined behaviour, including potential crashes ).

In the absence of literature, here's my own analysis:

First, consider the case of the a theoretically ideal 128 bit hash, and using a 128 bit sum to combine them.  The ideal hash will a number of desirable properties:  it is uniformly distributed over the 128 bit integers, different inputs produce uncorrelated results, changing any bit of the input will affect every bit of the output with a chance of 50%.   Summing an integer drawn uniformly from the 128 bit integers with another uncorrelated value produces a result distributed uniformly over the 128 bit integers ( assuming overflow wraps, as in C++ ).  Since all our element hashes include a unique name, then under this assumption of an ideal hash, all element hashes are uncorrelated, and this should be fine, and this applies recursively to sums of more than 2 elements.  Similarly, for either operand to the sum, each input bit pattern corresponds to a unique output bit pattern, the sum should preserve the property that changing any bit of either operand will affect every bit of the output with a chance of 50%.  Actually, thinking about this now, it might actually be easier to reason about this using xor instead of a sum - in general, xor is a poor choice for combining hashes, because combining two identical hashes would cancel out, but because our elements are all uncorrelated due to including the hash of their unique names, it seems like xor would also be just fine.  In this theoretical case, I'm fairly confident that using the sum does not increase the chances of collision.

Our practical case has two weaknesses compared to this theoretical one: we use 2 64-bit sums instead of 1 128-bit sum, and we use the actual hash MurmurHash, rather than a theoretically ideal hash.

The first concern I do not think is actually a weakness at all - based on my reasoning above, due to our operands already being uncorrelated, reducing the size of the sum modulos is not actually a concern.  Even using 128 separate modulo-2 sums ( as in xor ), would produce uniform results.  The only problem with 2 64-bit sums would be when dealing with correlated values, which with an ideal hash would only occur if the inputs were identical.  ( This is actually a concern with the first place I used this, the hash of the variations plug on Instancer, where a large number of identical inputs are expected.  My reasoning at the time was that you would only cancel out an element if you summed it 2**64 times, which seemed unlikely, but a better analysis is that you lose bits of entropy when summing an identical element - with a 128 bit sum, you would only end up with 128 - n bits of entropy if you sum 2**n identical elements.  With two 64-bit sums, you only end up with 128 - 2n bits of entropy when summing 2**n identical elements.  Not a huge concern in that case, since we still mix in other entropy sources, so the possibility of collision is only within the variations plug, which is only provided as a UI feedback, and a collision of two variation calculations would not be fatal ).  But anyway, in this case, where our inputs are uncorrelated, not a concern.

The second weakness is where the bulk of my concern lies:  MurmurHash is not a theoretically ideal hash.   We already rely on it being close to theoretically ideal, but it's difficult to rule out the possibility that there is some way it deviates from a perfectly ideal hash which is exposed through summing.  I do not think this is likely, but without detailed mathematical analysis of the MurmurHash algorithm in this context, it's hard to guarantee.

Actually, while writing this, I've thought of a test that would actually reassure me on this:  hash collisions are so unlikely that it seems difficult to test for them, but under the assumption that our hash is close to ideal, all parts of the hash should be equally well distributed.  We should be able to check the rate of collisions for a small portion of the hash ( ie. the lower 32 bits ), and if everything is working properly, this hash collision rate should not be affected by switching from an N-element hash to a sum of N 1-element hashes.  That definitely seems like a worthwhile test.

themissingcow commented:
Thanks for the time and detailed thoughts Daniel.

I have to confess that my maths brain operates on a far more basic level. My Q came from an incredibly naive (and most likely misguided) thought process based on imagining a simple 4 bit hash. I may be misunderstanding how all this works though, so apologies for the noise if this is the case.

Hash a context with a single entry, and it may give you 15.  There then seemed to then be multiple two entry contexts whose hash sum was also 15, even before considering overflows.

I didn't know if this was a correct statement. Did I understand your reasoning to be that, in the ideal hash case, this collision isn't actually much different from the chances of collision from hashing in the two entries in the normal way?

danieldresser-ie commented:
I think I would summarize your example with 4 bits, comparing to a single entry with a hash of 15, as follows:

If you individually sum the hashes of two entries, there are 16 possible the hashes of each entry, for a total of 256 possibilities.  Of these 256 possible sums, 16 of them are equal to 15: ( 0 + 15, 1 + 14, .... , 15 + 0 ).  So, there is a 16 / 256 chance of colliding with single element with a hash of 15, or 1 in 16 ( assuming the two individual hashes are uniformly distributed and indepedent ).

If we instead consider an ideal 4 bit hash of these two elements, it can only output one of the 16 4-bit values, so the chance of a collision is ... still 1 in 16.

Combining two hashes with a sum has multiple source values that will produce the same outputs ... but any theoretical method for combining two N-bit hashes into one N-bit hash has 2**(2N) possible inputs, and only 2**N outputs, so there will be on average 2**N possible input pairs that collide on any given output.  As long as the inputs are independent and uniformly distributed, I don't see any evidence that a sum is performing worse in this regard.

johnhaddon commented:
> V3: store the hash of each entry. To compute the current hash, run through all entries and sum their hashes

Question : is it worth trying "V4: store the hash of each entry. To compute the current hash, run through all entries and append to a final MurmurHash as before"?

danieldresser-ie commented:
I've now added your proposed V4, and updated the performance results at the top of this PR to show all 4 variants plus the original, run on both a smallish context and a largish one.

I've also added a test for the uniformity of the hash, which appears to confirm our belief that MurmurHash is performing close to ideally, and that changing the hash combining approach does not change this.

It seems like we've now got everything we need to either make a call on this ( I still like V3 ), or start thinking about storage approaches if we think the storage approach is going to interact enough that we need to think about that first.

danieldresser-ie commented:
I should mention one last minute catch - I was thinking this was basically good based on having run GafferTest, ImageTest, and SceneTest, but I finally checked why the interactive render tests were failing, and it's a legitimate bug:  I was calling the changedSignal before updating the hash values.  Obvious mistake, but I should probably add a test into GafferTest.ContextTest for it.

johnhaddon commented:
> It seems like we've now got everything we need to either make a call on this ( I still like V3 ), or start thinking about storage approaches if we think the storage approach is going to interact enough that we need to think about that first.

Thanks for going the extra mile on the testing and with V4. V3 gets my vote as well, and I think it gives us the best shot at improving the approach to storage. Could you clean up the PR to include just that approach, ready to merge?

We're also getting this crash in the debug tests, apparently related to the validation during destruction. Perhaps we don't need this now we're not going to be doing any subtract-then-add trickery?

```
* backtrace:
>> 0 0x00007f0896c04387 [libc.so.6                  ] __GI_raise                                                                                                                       [                     :              ?]
*  1 0x00007f0896c05a77 [libc.so.6                  ] __GI_abort                                                                                                                       [                     :              ?]
*  2 0x00007f0892fa47d4 [libstdc++.so.6             ] __gnu_cxx::__verbose_terminate_handler()                                                                                         
*  3 0x00007f0892fa2745 [libstdc++.so.6             ] std::rethrow_exception(std::__exception_ptr::exception_ptr)                                                                      
*  4 0x00007f0892fa16f8 [libstdc++.so.6             ] __cxa_bad_typeid                                                                                                                 
*  5 0x00007f0892fa2363 [libstdc++.so.6             ] __gxx_personality_v0                                                                                                             
*  6 0x00007f0892d3d8a2 [libgcc_s.so.1              ] _Unwind_GetTextRelBase                                                                                                           
*  7 0x00007f0892d3ddd6 [libgcc_s.so.1              ] _Unwind_Resume                                                                                                                   
*  8 0x00007f088957d9bb [libGaffer.so               ] Gaffer::Context::validateHashes()                                                                                                [Context.cpp          :491 (discrimina]
*  9 0x00007f088957bc56 [libGaffer.so               ] Gaffer::Context::~Context()
```

danieldresser-ie commented:
OK, I guess I can call this ready for review now.  Definitely still needs some close inspection, but should be a lot easier now that we're just focussing on V3.

Some things to think about:
> We're also getting this crash in the debug tests, apparently related to the validation during destruction. Perhaps we don't need this now we're not going to be doing any subtract-then-add trickery?

There is still the potential for the per-element hash to become desynchronized from it's element value, so I think this is still worthwhile in debug mode?  In this case, it seems like a pretty legitimate failure was identified:  GafferSceneUI::ContextAlgo changes context entries without calling Context::set, and then instead of calling Context::changed(), it manually triggered Context::changedSignal() ... I'm not sure if there was a reason for this.

Also, my test that hash collisions occur at an expected rate got pretty expensive after I got a bit more paranoid and decided to check each of the 32 bit chunks of the hash to make sure they are all uniform.  Maybe I need to turn it off by default somehow ( it's already deactivated on CI )?

danieldresser-ie commented:
The remaining test failure I'm seeing is a bit of a weird one:  I've now corrected GafferSceneUI::ContextAlgo::expandDescendants so that after changing the Context it calls Context::changed, but SceneView::expandSelection calls expandDescendants on a Context that is currently scoped, so while it is in the middle of being modified, it is being used in a traversal, and a call to childNames() creates a temp context for a scope which is initialiazed from the parent context that is in an in-between state, and on destruct in debug mode, it notices that the per-entry hash is inconsistent.  The reason this is not an actual problem is because the per-entry hash for the ui:scene:expandedPaths is not actually used, but putting an context in an inconsistent state while it's in use does still seem legitimately off.

johnhaddon commented on src/GafferTest/ContextTest.cpp:
```diff
@@ -204,3 +205,39 @@ void GafferTest::testEditableScope()
+	ContextPtr baseContext = new Context();
+	for( int i = 0; i < numEntries; i++ )
+	{
+		baseContext->set( std::to_string( i ), std::string( entrySize, 'x') );
```
Probably not important here, but there's an `InternedString( int64_t )` constructor that should be faster than `InternedString( to_string( i ) )`.

danieldresser-ie replied:
Oh, good point.

danieldresser-ie replied:
Fixed in 1820604

johnhaddon commented on src/GafferTest/ContextTest.cpp:
```diff
@@ -204,3 +205,39 @@ void GafferTest::testEditableScope()
+
+	tbb::parallel_for( tbb::blocked_range<size_t>( 0, 10000000 ), [&threadState, &varyingVarName]( const tbb::blocked_range<size_t> &r )
+		{
+			// As part of the setCollaborate plug machinery, we put the parentPath in the context.
+			// Need to remove it before evaluating the prototype sets
```
Comment seems unrelated.

danieldresser-ie replied:
Oops.

danieldresser-ie replied:
Fixed in 717b9a8

johnhaddon commented on python/GafferTest/ContextTest.py:
```diff
@@ -684,6 +684,66 @@ def testOmitCanceller( self ) :
+	def collisionCountParallelHelper( seed, mode, chunk, countList ):
+		countList[0] = GafferTest.countContextHash32Collisions( 2**20, mode, chunk, seed )
+
+	@unittest.skipIf( GafferTest.inCI(), "Too expensive to run on CI, and has a small chance of failure because it is statistical" )
```
This is too expensive to run as a matter of course. I rely on being able to run the full test suite for a module fairly regularly as I work, and this kills that ability (a few recent additions are pushing the boundaries already). Perhaps we should extend the `gaffer test -performanceOnly` idea to something like `gaffer test -type all|standard|performance|veryVerySlow`?

johnhaddon replied:
> I guess the quick variant would be setting skipIf( True ) and leaving it up to anyone who is doing hash experimentation in the future to manually turn it on?

I'll buy that for now.

> If we do take a look at the args to gaffer test, one that I should add is "always print performance times".

Sounds good. Let's aim to tackle these together in the not-too-distant future.

johnhaddon commented on src/GafferTest/ContextTest.cpp:
```diff
@@ -206,6 +207,67 @@ void GafferTest::testEditableScope()
+{
+	std::unordered_set<uint32_t> used;
+
+	srand( seed );
```
This isn't thread-safe, and we do seem to be calling this function multiple times in parallel. Perhaps `rand_r()` or something out of `https://en.cppreference.com/w/cpp/numeric/random`.

danieldresser-ie replied:
Oh, good point.

danieldresser-ie replied:
Turned out to be a very good point: in combination with the precreation of InternedStrings mentioned below, switching to rand_r() means this test is now only taking 70 seconds, and the distribution does not seem to have been affected.

danieldresser-ie replied:
Fixed in 717b9a8

johnhaddon commented on src/GafferTest/ContextTest.cpp:
```diff
@@ -206,6 +207,67 @@ void GafferTest::testEditableScope()
+		Context c;
+		if( curMode == 0 )
+		{
+			c.set( "a", element );
```
Since this test is slow, it's probably worth pulling the InternedString construction out of the inner loop.

danieldresser-ie replied:
Good catch - reduced time from 500 seconds to 300 seconds.  Still too slow to leave turned on, but certainly doesn't hurt.

danieldresser-ie replied:
Fixed in 717b9a8

johnhaddon commented on src/GafferTest/ContextTest.cpp:
```diff
@@ -206,6 +207,67 @@ void GafferTest::testEditableScope()
+			{
+				c.set( std::to_string(j), j );
+			}
+			c.set( "a", element );
```
As above.

johnhaddon commented on src/GafferTest/ContextTest.cpp:
```diff
@@ -206,6 +207,67 @@ void GafferTest::testEditableScope()
+		{
+			for( int j = 0; j < 40; j++ )
+			{
+				c.set( std::to_string(j), j );
```
Probably worth using the `InternedString( int64_t )` constructor here, to try to get the cost of this test down.

johnhaddon commented on python/GafferTest/ContextTest.py:
```diff
@@ -684,6 +684,66 @@ def testOmitCanceller( self ) :
+				for l in results:
+					s += l[0]
+
+				# countContextHash32Collisions looks at just the lowest 32 bits of the hash.
```
I thought we are looking at each of the four 32 bit chunks?

danieldresser-ie replied:
Oh, yep, stale comment.

danieldresser-ie replied:
Fixed in 717b9a8

johnhaddon commented on src/GafferTest/ContextTest.cpp:
```diff
@@ -206,6 +207,67 @@ void GafferTest::testEditableScope()
+	for( int i = 0; i < contexts; i++ )
+	{
+		int curMode = mode;
+		int element = seed * contexts + i;
```
This may be a stupid question, but I assume `element` is the "entry with the single increment int", so why the `seed * contexts` constant offset? To try to keep the values used completely different for each seed?

danieldresser-ie replied:
I should probably name it "elementSeed", and yes, the goal is to keep the values used completely different for each seed.  We run this test multiple times to reduce random variation, so that we can expect to see the ideal number of collisions +/- 10%, rather than +/- 50%.  If we used the same values for the different runs, we wouldn't get this variation reduction.

johnhaddon commented on src/GafferTest/ContextTest.cpp:
```diff
@@ -206,6 +207,67 @@ void GafferTest::testEditableScope()
+		{
+			for( int j = 0; j < 20; j++ )
+			{
+				c.set( std::to_string(j), rand() );
```
As above.

johnhaddon commented on src/Gaffer/Context.cpp:
```diff
@@ -477,3 +480,17 @@ const std::string &Context::SubstitutionProvider::variable( const boost::string_
+{
+	for( Map::iterator it = m_map.begin(), eIt = m_map.end(); it != eIt; ++it )
+	{
+		IECore::MurmurHash correct = it->second.data->Object::hash();
+		correct.append( (uint64_t)&it->first.string() );
```
This is repeated in three spots - might be worth putting in a new `Storage::updateHash( name )` method?

danieldresser-ie replied:
Sure - in a way it would be nice if it was a private method on Context, so it could set `m_hashValid = false` as well, but it's probably cleaner to put it on Storage and pair it with a `m_hashValid = false`

danieldresser-ie replied:
Fixed in 7e95c0f

johnhaddon commented on src/GafferSceneUI/ContextAlgo.cpp:
```diff
@@ -148,9 +148,9 @@ void expand( Context *context, const PathMatcher &paths, bool expandAncestors )
-		// so the context doesn't know they've changed. So we emit the changed
-		// signal ourselves
-		context->changedSignal()( context, g_expandedPathsName );
+		// so the context doesn't know they've changed. So we must let it know
+		// about the change
```
I think the previous comment just indicates that I'd overlooked the need to call `changed()` rather than emit `changedSignal()`. I don't think I was trying to do anything clever.

johnhaddon reviewed (changes requested):
Thanks Daniel! Great to see such substantial improvements to the performance of a core component. Comments inline as usual.

> The remaining test failure I'm seeing is a bit of a weird one

Ugh, yeah, that is weird. I really should never have used the script's context to store state between UI components in the first place. I made one comment inline about storing an empty hash for `ui:*`, so perhaps that might sweep it back under the rug for a bit? Really I think we need to move the storage of all that stuff into a completely different object.

danieldresser-ie commented on include/Gaffer/Context.h:
```diff
@@ -275,14 +275,16 @@ class GAFFER_API Context : public IECore::RefCounted
 
 		Map m_map;
 		ChangedSignal *m_changedSignal;
-		mutable IECore::MurmurHash m_hash;
-		mutable bool m_hashValid;
```
Yeah, you're totally right - these were ripped out for V1, where they really weren't necessary any more, but for the approach we've decided on, they probably are needed again.

danieldresser-ie replied:
One weird corner case I hesitated on ... in the previous code, calling `changed( "variableNameThatDoesntExist" )` would set m_hashValid to false.  Now that I'm actually finding the entry for the hash value ( so that I can recompute the per-entry hash ), it feels more natural and correct to only invalidate m_hashValid if the name refers to a variable that actually exists in the context?  Any opinions on this change in behaviour?

danieldresser-ie replied:
Fixed in 7e95c0f

johnhaddon replied:
> it feels more natural and correct to only invalidate m_hashValid if the name refers to a variable that actually exists in the context? Any opinions on this change in behaviour?

Yep, sounds good.

danieldresser-ie commented on include/Gaffer/Context.inl:
```diff
@@ -181,12 +181,16 @@ void Context::set( const IECore::InternedString &name, const T &value )
-		m_hashValid = false;
+		const std::string &nameStr = name.string();
+		s.hash = s.data->Object::hash();
+		s.hash.append( (uint64_t)&nameStr );
```
I can try it ... I suspect it would be a measurable loss of performance?  I kinda wish there was a unique integer associated with InternedStrings based on insertion order, so that we could use that, and at least get the possibility of determinism for the tests if we pre-inserted the strings.  Though, actually, if we were going to do anything to InternedStrings to make them better for use in Gaffer, it would be tempting to store hashes with the central InternedString ... it seems pretty rare that we create a string in Gaffer and don't need to hash it at some point.

johnhaddon replied:
Happy to merge as we are and then do the experiment after. Storing a hash with the InternedString itself sounds interesting - I wonder how much of the cost associated with hashing the string itself is MurmurHash and how much is just the memory access. Another experiment for after the merge perhaps?

danieldresser-ie commented on src/Gaffer/Context.cpp:
```diff
@@ -304,15 +307,15 @@ IECore::MurmurHash Context::hash() const
 		// it's quicker than `string::compare( 0, 3, "ui:" )`.
 		const std::string &name = it->first.string();
-		if(	name.size() > 2 && name[0] == 'u' && name[1] == 'i' && name[2] == ':' )
+		if( name.size() > 2 && name[0] == 'u' && name[1] == 'i' && name[2] == ':' )
```
Oh, that's totally fair.  I was eager to get it out of `set()` because V1 required this logic to match in `set`, `remove` and `removeMatching`, which felt horrifically fragile.  But now that we don't care about removal, putting it in `set` is actually fairly nice.

danieldresser-ie replied:
Put it in the new updateHash function on Storage, which feels pretty nice now.

danieldresser-ie replied:
Fixed in 7e95c0f

danieldresser-ie commented on Changes.md:
```diff
@@ -23,6 +23,7 @@ Improvements
 - Improved performance of Python expressions when they are evaluated multiple times in parallel. This is the first use of the Standard cache policy in a node which could be downstream of a custom Gaffer node that could make an improperly isolated call to tbb, triggering a hang. If this causes a hang, set the env var GAFFER_PYTHONEXPRESSION_CACHEPOLICY=Legacy as a temporary workaround, and make su ...
 - Increased image processing tile size from 64 pixels to 128 pixels. This reduces per-tile overhead on large images, dramatically increasing effective image performance in many cases.
 - OSLImage : Avoided some unnecessary computes and hashing when calculating channel names or passing through channel data unaltered.
+- Context : Optimized hash() method
```
Fixed in 7e95c0f

danieldresser-ie commented on include/Gaffer/Context.h:
```diff
@@ -275,14 +275,16 @@ class GAFFER_API Context : public IECore::RefCounted
 			// And use this ownership flag to tell us when we need to do explicit
 			// reference count management.
 			Ownership ownership;
+			// This tells us the hash value we need to subtract off the incremental hash if we change this value
```
Fixed in 7e95c0f

danieldresser-ie commented:
In addition to addressing comments, I've also added the commit "SceneView : Don't scope Context while we're modifying it"

johnhaddon commented on src/GafferSceneUI/SceneView.cpp:
```diff
@@ -1810,7 +1810,10 @@ void SceneView::frame( const PathMatcher &filter, const Imath::V3f &direction )
-	Context::Scope scope( getContext() );
+	// Create a copy of the current context to evaluate the scene in, so that we aren't using the
+	// current context while we're modifying it.
+	Context referenceContext( *getContext(), Context::Copied );
```
This should be `ContextPtr .. = new Context()`. Constructing a RefCounted object on the stack is asking for trouble, because it'll get double destructed if something else decides to add a reference (for instance, to pass it to Python).

I'm a little bit worried about the performance consequences here though, because it'll cause a copy-on-write for the modifications we subsequently make to the expanded paths. Let's chat about that in our catch up today.

johnhaddon commented on src/GafferTest/ContextTest.cpp:
```diff
@@ -204,3 +206,106 @@ void GafferTest::testEditableScope()
 
 }
+
+// Create the number of contexts specified, and return a count for how many collisions there are
+// in the last 32 bits of the context hash.  The uniformity of the last 32 bits should be
+// representative of the uniformity of the hash as a whole ( we don't want to check for collisions
+// in the whole 128 bit hash, since it would take years for one to occur randomly )
```
This is still talking about the low 32 bits rather than the chunk.

johnhaddon reviewed (changes requested):
GitHub's done that annoying thing where it posts the review early if you also comment on a previous conversation. So there's one comment here and one dangling one above...

Did you get a chance to get new numbers after removing the `ui:*` check out of the main hashing loop? Would be interesting to know if it had any impact...

danieldresser-ie commented:
> Did you get a chance to get new numbers after removing the ui:* check out of the main hashing loop?

I think there is some measurable effect, but it's smaller than the amount of noise in the test, so it's hard to measure precisely.  Something like a 0.005 improvement out of 0.35

Your comment about the inconsistent comment in the ContextTest collision code made me realize that if we're consistently checking all four 32 bit chunks of the hash, then countContextHash32Collisions should just return four counts, one for each chunk, instead of having a parameter to say which chunk you want.  This sped up the test from ~70 seconds to ~30 seconds.

I've squashed all fixes, and reverted the change to SceneView - instead, I just add a comment pointing out that what it's doing is a little weird.

johnhaddon commented:
Thanks Daniel!
